pwcnet/model.py

```diff
@@ -0,0 +1,157 @@
+        self.num_feature_levels = 6
+        # This is the output feature level. This can be anywhere between [1, self.num_feature_levels].
+        self.output_level = 2
+        # A range that counts down from [self.num_flow_estimates, self.output_level] inclusive.
+        self.iter_range = range(self.num_feature_levels, self.output_level - 1, -1)
+
+        self.feature_pyramid = FeaturePyramidNetwork(regularizer=self.regularizer)
+        self.estimator_networks = [EstimatorNetwork(name='estimator_network_' + str(i), regularizer=self.regularizer)
+                                   for i in self.iter_range]
+        self.context_network = ContextNetwork(regularizer=self.regularizer)
+
+    def get_forward(self, image_a, image_b):
+        """
+        :param image_a: Tensor of shape [batch_size, H, W, 3].
+        :param image_b: Tensor of shape [batch_size, H, W, 3].
+        :return: final_flow: up-sampled final flow.
+                 previous_flows: all previous flow outputs of the estimator networks and the context network.
+        """
+        with tf.variable_scope(self.name):
+            img_height = tf.shape(image_a)[1]
+            img_width = tf.shape(image_a)[2]
+            # Siamese networks (i.e. image_a and image_b are fed through the same network with shared weights).
+            _, features_a = self.feature_pyramid.get_forward(image_a, reuse_variables=tf.AUTO_REUSE)
+            _, features_b = self.feature_pyramid.get_forward(image_b, reuse_variables=tf.AUTO_REUSE)
+
+            # The initial flow will be initialized with zeros.
+            # It is refined at each feature level.
+            previous_flow = None
+            previous_flows = []
+
+            # Counts down from [self.num_flow_estimates, self.output_level] inclusive.
+            for i in self.iter_range:
+                if VERBOSE:
+                    print('Creating estimator at level', i)
+                # Get the features at this level.
+                features_a_n = features_a[self.feature_pyramid.get_c_n_idx(i)]
+                features_b_n = features_b[self.feature_pyramid.get_c_n_idx(i)]
+
+                # Setup the previous flow for input into the estimator network at this level.
+                B = tf.shape(features_a_n)[0]
+                H = tf.shape(features_a_n)[1]
+                W = tf.shape(features_a_n)[2]
+                if previous_flow is None:
+                    previous_flow = tf.zeros(shape=[B, H, W, 2], dtype=tf.float32)
+                else:
+                    # The original scale flows at all layers is the same as the scale of the ground truth.
+                    scaling = tf.cast(H, tf.float32) / tf.cast(img_height, tf.float32)
+                    previous_flow = previous_flow * scaling / self.flow_scaling
+                    # Upsample to the size of the current layer.
+                    previous_flow = tf.image.resize_images(previous_flow, [H, W],
```

Comment:
I think that this code makes sense, but just to double check, does it correspond with what the [paper](https://arxiv.org/pdf/1709.02371.pdf) is saying in their Experimental Results section? It seems that they scaled their ground truth flow by 20, maybe it was miscommunicated?

Reply:
Yea so I got confused by how they worded it too. When they said "scaled by 20", I think they meant divided by 20. Cus in an example like 2 sentences later, they wrote that they scaled the intermediate flows by 20/(2^layer), which only makes sense if they divided the ground truth by 20. Also, dividing the ground truth by 20 makes kinda sense since that means they're assuming a stddev of 20 and "normalizing" it in a way.